micro-rdk/src/esp32/pin.rs

```diff
@@ -38,7 +38,7 @@ fn install_gpio_isr_service() -> Result<(), BoardError> {
 // ESP-IDF, so we must replicate the logic from that function here. If we do not validate,
 // PinDriver::input_output will panic because esp-idf-hal does not perform the same check
 fn is_valid_gpio_pin(pin: i32) -> Result<(), BoardError> {
-    match (1_u32 << (pin as u32)) & SOC_GPIO_VALID_DIGITAL_IO_PAD_MASK {
+    match (1_u32 << (pin as u32)) & (SOC_GPIO_VALID_DIGITAL_IO_PAD_MASK as u32) {
```

Comment:
This confused me. For esp32s3 it was acting as if this macro was a u64 but I don't see how that could be. @gvaradarajan definitely interested if you have any insights since you added this recently.

Reply:
how did you observe it was acting as a u64?

Reply:
Without the cast, I get the following on esp32s3 targets:
```
error: could not compile `micro-rdk` (lib) due to 3 previous errors
error[E0308]: mismatched types
  --> micro-rdk/src/esp32/pin.rs:41:37
   |
41 |     match (1_u32 << (pin as u32)) & SOC_GPIO_VALID_DIGITAL_IO_PAD_MASK {
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u32`, found `u64`

error[E0277]: no implementation for `u32 & u64`
  --> micro-rdk/src/esp32/pin.rs:41:35
   |
41 |     match (1_u32 << (pin as u32)) & SOC_GPIO_VALID_DIGITAL_IO_PAD_MASK {
   |                                   ^ no implementation for `u32 & u64`
   |
   = help: the trait `BitAnd<u64>` is not implemented for `u32`
   = help: the following other types implement trait `BitAnd<Rhs>`:
             <u32 as BitAnd>
             <u32 as BitAnd<&u32>>
             <&'a u32 as BitAnd<u32>>
             <&u32 as BitAnd<&u32>>

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
```

Unfortunately, *with* the cast I get clippy complaints on plain esp32, per CI. See https://github.com/viamrobotics/micro-rdk/actions/runs/10219337838/job/28277453866?pr=288#step:6:815.

Though probably easier to just silence clippy in that case.

Reply:
Found it.

That constant comes from esp-idf, and it has different implementations depending on the target.
So you actually don't want to be casting here as you'd be missing out on the rest of the mask.

[esp32](https://github.com/espressif/esp-idf/blob/41dd1a351b4f8630a0b9dbdd87a6bf782094dd66/components/soc/esp32/include/soc/soc_caps.h#L190): `#define SOC_GPIO_VALID_DIGITAL_IO_PAD_MASK 0xEF0FEAULL`

[esp32s3](https://github.com/espressif/esp-idf/blob/41dd1a351b4f8630a0b9dbdd87a6bf782094dd66/components/soc/esp32s3/include/soc/soc_caps.h#L187) : `#define SOC_GPIO_VALID_DIGITAL_IO_PAD_MASK 0x0001FFFFFC000000ULL`

Reply:
@mattjperez - Nice find, thanks for that. I'm going to try to figure out a way that works for both targets.